lib/resty/jwt-validators.lua

```diff
@@ -0,0 +1,384 @@
+  ensure_is_type(check_val, "number", messages.wrong_type_validator, "number", "check_val")
+  return _M.opt_check(check_val, less_than_or_equal_function, "check_val", "number")
+end)
+
+
+--[[
+    A function to set the leeway (in seconds) used for is_not_before and is_not_expired.  The
+    default is to use 0 seconds
+]]--
+local system_leeway = 0
+function _M.set_system_leeway(leeway)
+  ensure_is_type(leeway, "number", "leeway must be a non-negative number")
+  ensure_is_non_negative(leeway, "leeway must be a non-negative number")
+  system_leeway = leeway
+end
+
+
+--[[
+    A function to set the system clock used for is_not_before and is_not_expired.  The
+    default is to use ngx.now
+]]--
+local system_clock = ngx.now
+function _M.set_system_clock(clock)
+  ensure_is_type(clock, "function", "clock must be a function")
+  
+  -- Check that clock returns the correct value
+  local t = clock()
+  ensure_is_type(t, "number", "clock function must return a non-negative number")
+  ensure_is_non_negative(t, "clock function must return a non-negative number")
+  system_clock = clock
+end
+
+-- Local helper function for date validation
+local function validate_is_date(val, claim, jwt_json)
+  ensure_is_non_negative(val, messages.wrong_type_claim, claim, "positive numeric value")
+  return true
+end
+
+-- Local helper for date formatting
+local function format_date_on_error(date_check_function, error_msg)
+  ensure_is_type(date_check_function, "function", messages.wrong_type_validator, "function", "date_check_function")
+  ensure_is_type(error_msg, "string", messages.wrong_type_validator, "string", error_msg)
+  return function(val, claim, jwt_json)
+    local ret = date_check_function(val, claim, jwt_json)
+    if ret == false then
+      error(string.format("'%s' claim %s %s", claim, error_msg, ngx.http_time(val)))
+    end
+    return true
+  end
+end
+
+--[[
+    Returns a validator that checks if the current time is not before the tested value
+    within the system's leeway.  This means that:
+      val <= (system_clock() + system_leeway).
+]]--
+define_validator("is_not_before", function()
+  return format_date_on_error(
+    _M.chain(validate_is_date, _M.opt_less_than_or_equal(system_clock() + system_leeway)),
+    "not valid until"
+  )
+end)
+
+
+--[[
+    Returns a validator that checks if the current time is not equal to or after the 
+    tested value within the system's leeway.  This means that:
+      val > (system_clock() - system_leeway).
+]]--
+define_validator("is_not_expired", function()
+  return format_date_on_error(
+    _M.chain(validate_is_date, _M.opt_greater_than(system_clock() - system_leeway)),
+    "expired at"
+  )
+end)
+
+
+--[[
+    Returns a validator that checks if the current time is the same as the tested value 
+    within the system's leeway.  This means that:
+      val >= (system_clock() - system_leeway) and val <= (system_clock() + system_leeway).
+]]--
+define_validator("is_at", function()
```

Comment:
Hmmm. I must be missing something.

In order to check the `iat` claim, I'd may check its value isn't greater than `now` (+/- leeway). Optionally I may enforce it's less than `exp` (+/- leeway), in order to assess the jwt integrity.

However, I'm not sure how checking `iat` equals `now` could be any reliable. A jwt is supposed to last for some time, and be reused across many calls, so as soon as the leeway is exceeded, this validation will be failing (even if the jwt is still valid (eg. not expired yet)).

It's far possible I have misunderstood your point. Would that be the case, i apologize in advance. However, I'd be a taker for any explanation you may share to help me get a better grasp at this subject.